File: mobile/android/app/src/main/java/com/tigerwallet/app/services/options/OptionsTradingService.java
package com.tigerwallet.app.services;

import org.json.JSONArray;
import org.json.JSONObject;
import java.net.HttpURLConnection;
import java.net.URL;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;

public class OptionsTradingService {
    private static final String API_BASE = "https://api.tigerwallet.com/api/v1";
    private String authToken;
    
    public OptionsTradingService(String token) { this.authToken = token; }
    
    private HttpURLConnection createConnection(String endpoint) throws Exception {
        URL url = new URL(API_BASE + endpoint);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("Content-Type", "application/json");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        return conn;
    }
    
    public OptionsContract[] getOptions() throws Exception {
        HttpURLConnection conn = createConnection("/options/contracts");
        if (conn.getResponseCode() == 200) {
            String response = readResponse(conn);
            JSONArray data = new JSONObject(response).getJSONArray("data");
            OptionsContract[] contracts = new OptionsContract[data.length()];
            for (int i = 0; i < data.length(); i++) {
                JSONObject c = data.getJSONObject(i);
                contracts[i] = new OptionsContract(c.getString("id"), c.getString("symbol"), 
                    c.getDouble("strike"), c.getDouble("premium"), c.getString("expiry"), c.getString("type"));
            }
            return contracts;
        }
        return new OptionsContract[0];
    }
    
    public OptionsOrder buyOption(String contractId, int quantity) throws Exception {
        URL url = new URL(API_BASE + "/options/buy");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        conn.setDoOutput(true);
        JSONObject body = new JSONObject();
        body.put("contractId", contractId);
        body.put("quantity", quantity);
        conn.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8));
        
        if (conn.getResponseCode() == 201) {
            String response = readResponse(conn);
            JSONObject data = new JSONObject(response).getJSONObject("data");
            return new OptionsOrder(data.getString("id"), data.getString("contractId"), 
                data.getDouble("quantity"), data.getDouble("price"), data.getString("status"));
        }
        return null;
    }
    
    public boolean exerciseOption(String orderId) throws Exception {
        URL url = new URL(API_BASE + "/options/" + orderId + "/exercise");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        return conn.getResponseCode() == 200;
    }
    
    private String readResponse(HttpURLConnection conn) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) response.append(line);
        reader.close();
        return response.toString();
    }
    
    public static class OptionsContract {
        public String id, symbol, expiry, type;
        public double strike, premium;
        public OptionsContract(String id, String symbol, double strike, double premium, String expiry, String type) {
            this.id = id; this.symbol = symbol; this.strike = strike; this.premium = premium;
            this.expiry = expiry; this.type = type;
        }
    }
    
    public static class OptionsOrder {
        public String id, contractId, status;
        public double quantity, price;
        public OptionsOrder(String id, String contractId, double quantity, double price, String status) {
            this.id = id; this.contractId = contractId; this.quantity = quantity; this.price = price; this.status = status;
        }
    }
}

File: mobile_apps/ios_legacy/Runner/Services/options/OptionsTradingService.swift
import Foundation

class OptionsTradingService {
    static let API_BASE = "http://localhost:8443/api/v1"
    var authToken: String?
    
    init(token: String?) { self.authToken = token }
    
    private var headers: [String: String] {
        var h = ["Content-Type": "application/json"]
        if let token = authToken { h["Authorization"] = "Bearer \(token)" }
        return h
    }
    
    func getOptions(completion: @escaping ([OptionContract]) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/options/contracts") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.allHTTPHeaderFields = headers
        
        URLSession.shared.dataTask(with: request) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let array = json["data"] as? [[String: Any]] else {
                completion([])
                return
            }
            let contracts = array.compactMap { dict -> OptionContract? in
                guard let id = dict["id"] as? String, let symbol = dict["symbol"] as? String else { return nil }
                return OptionContract(id: id, symbol: symbol, strike: dict["strike"] as? Double ?? 0, premium: dict["premium"] as? Double ?? 0)
            }
            completion(contracts)
        }.resume()
    }
    
    func buyOption(contractId: String, quantity: Int, completion: @escaping (Bool) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/options/buy") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.allHTTPHeaderFields = headers
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["contractId": contractId, "quantity": quantity])
        
        URLSession.shared.dataTask(with: request) { _, response, _ in
            if let http = response as? HTTPURLResponse { completion(http.statusCode == 201) }
            else { completion(false) }
        }.resume()
    }
}

struct OptionContract {
    let id: String
    let symbol: String
    let strike: Double
    let premium: Double
}

class CopyTradingService {
    static let API_BASE = "http://localhost:8443/api/v1"
    var authToken: String?
    
    init(token: String?) { self.authToken = token }
    
    private var headers: [String: String] {
        var h = ["Content-Type": "application/json"]
        if let token = authToken { h["Authorization"] = "Bearer \(token)" }
        return h
    }
    
    func getTopTraders(completion: @escaping ([Trader]) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/copy/traders") else { return }
        var request = URLRequest(url: url)
        request.allHTTPHeaderFields = headers
        
        URLSession.shared.dataTask(with: request) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let array = json["data"] as? [[String: Any]] else {
                completion([])
                return
            }
            let traders = array.compactMap { dict -> Trader? in
                guard let id = dict["id"] as? String, let name = dict["name"] as? String else { return nil }
                return Trader(id: id, name: name, pnl: dict["pnl"] as? Double ?? 0, followers: dict["followers"] as? Double ?? 0)
            }
            completion(traders)
        }.resume()
    }
    
    func followTrader(traderId: String, amount: Double, completion: @escaping (Bool) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/copy/follow") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.allHTTPHeaderFields = headers
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["traderId": traderId, "amount": amount])
        
        URLSession.shared.dataTask(with: request) { _, response, _ in
            if let http = response as? HTTPURLResponse { completion(http.statusCode == 201) }
            else { completion(false) }
        }.resume()
    }
}

struct Trader {
    let id: String
    let name: String
    let pnl: Double
    let followers: Double
}
